File: 169.majority-element.js
/*
 * @lc app=leetcode id=169 lang=javascript
 *
 * [169] Majority Element
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    if(nums.length == 1) {
        return nums[0];
    }
    maxAppearances = Math.floor(nums.length / 2);
    totalAppearances = 0;
    nums.sort(function(a, b){return a - b});
    for(var i = 0; i < nums.length; i++) {
        if(nums[i] == nums[i+1]) {
            totalAppearances += 1;
            if(totalAppearances == maxAppearances) {
                return nums[i];
            }
        }
        else {
            totalAppearances = 0;
        }
    }
};
// @lc code=end

File: python3/169.majority-element.py
#
# @lc app=leetcode id=169 lang=python3
#
# [169] Majority Element
#


# @lc code=start
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        count = 0
        for num in nums:
            if count == 0:
                result = num
                count += 1
            elif num == result:
                count += 1
            else:
                count -= 1
        return result


# @lc code=end

# Boyer-Moore Voting Algorithm
# Time Complexity: O(n)
# Space Complexity: O(1)
